[EH] Troubleshooting Guide Updates (#38962)
Azure/azure-sdk-for-python: sdk/eventhub/azure-eventhub/TROUBLESHOOTING.md

- [Permission issues](#permission-issues)
- [Connectivity issues](#connectivity-issues)
  - [Timeout when connecting to service](#timeout-when-connecting-to-service)
    - [SSL handshake failures](#ssl-handshake-failures)
    - [Managing clients](#managing-clients)

>     - [Cannot add components to the connection string](#cannot-add-components-to-the-connection-string)

Reviewer: this one isn't working for me

Reviewer: cannot-add-connection-string

  - [Duplicate events are being processed](#duplicate-events-are-being-processed)
  - [Soft Delete or Blob versioning is enabled for a Blob Storage checkpoint store](#soft-delete-or-blob-versioning-is-enabled-for-a-blob-storage-checkpoint-store)
  - [Migrate from legacy to new client library](#migrate-from-legacy-to-new-client-library)
- [Get additional help](#get-additional-help)
  - [Filing GitHub issues](#filing-github-issues)

## Handle Event Hubs exceptions

All Event Hubs exceptions are wrapped in an [EventHubError][EventHubError].  They often have an underlying AMQP error code which specifies whether an error should be retried.  For retryable errors (ie. `amqp:connection:forced` or `amqp:link:detach-forced`), the client libraries will attempt to recover from these errors based on the retry options specified when instantiating the client:
* `retry_total`: The total number of attempts to redo a failed operation when an error occurs
* `retry_backoff_factor`: A backoff factor to apply between attempts after the second try
* `retry_backoff_max`: The maximum back off time
* `retry_mode`: The delay behavior between retry attempts. Supported values are 'fixed' or 'exponential'

> To configure retry options, follow the sample [client_retry][ClientCreation].  If the error is non-retryable, there is some configuration issue that needs to be resolved.

Author: good catch

The recommended way to solve the specific exception the AMQP exception represents is to follow the

> [Event Hubs Messaging Exceptions][EventHubsMessagingExceptions] guidance.

Reviewer: line 64 formatted wrong

- [Connectivity issues](#connectivity-issues)
  - [Timeout when connecting to service](#timeout-when-connecting-to-service)
  - [SSL handshake failures](#ssl-handshake-failures)
  - [Managing clients](#managing-clients)
  - [Adding components to the connection string does not work](#adding-components-to-the-connection-string-does-not-work)

>     - ["TransportType=AmqpWebSockets" Alternative](#transporttypeamqpwebsockets-alternative)

Reviewer 2: I think we should probably change ours to AmqpOverWebsocket, as per our naming

Reviewer 2: Both this line and the one below should be changed to match the section title so the link is clickable.

Reviewer 2: Same thing for a few other section titles

- [Connectivity issues](#connectivity-issues)
  - [Timeout when connecting to service](#timeout-when-connecting-to-service)
  - [SSL handshake failures](#ssl-handshake-failures)
  - [Managing clients](#managing-clients)
  - [Adding components to the connection string does not work](#adding-components-to-the-connection-string-does-not-work)
    - ["TransportType=AmqpOverWebsocket" Alternative](#adding-transporttypeamqpoverwebsocket)
    - ["Authentication=Managed Identity" Alternative](#adding-authenticationmanaged-identity)
  - [Connect using an IoT connection string](#connect-using-an-iot-connection-string)
  - [Cannot add components to the connection string](#cannot-add-components-to-the-connection-string)
- [Enable and configure logging](#enable-and-configure-logging)
  - [Enable AMQP transport logging](#enable-amqp-transport-logging)

> - [Troubleshoot EventProducerAsyncClient/EventProducerClient issues](#troubleshoot-eventproducerasyncclienteventproducerclient-issues)

Reviewer 2: link is not clickable

Reviewer: https://github.com/Azure/azure-sdk-for-python/blob/c66d45e3d2de2d158064d659b019fe26ac67ce9d/sdk/eventhub/azure-eventhub/TROUBLESHOOTING.md#**troubleshoot-eventhubproducerclient-syncasync-issues**

troubleshoot-eventhubproducerclient-syncasync-issues

  - [Cannot set multiple partition keys for events in EventDataBatch](#cannot-set-multiple-partition-keys-for-events-in-eventdatabatch)
  - [Setting partition key on EventData is not set in Kafka consumer](#setting-partition-key-on-eventdata-is-not-set-in-kafka-consumer)
  - [Buffer producer not sending events](#buffered-producer-not-sending-events)

> - [Troubleshoot EventHubConsumerClient issues](#troubleshoot-eventprocessorclient-issues)

Reviewer 2: link is not clickable

Author: this one is working for me locally ?

Reviewer: https://github.com/Azure/azure-sdk-for-python/blob/c66d45e3d2de2d158064d659b019fe26ac67ce9d/sdk/eventhub/azure-eventhub/TROUBLESHOOTING.md#troubleshoot-eventhubconsumerclient-issues

troubleshoot-eventhubconsumerclient-issues

This error can occur when an intercepting proxy is used.  We recommend testing in your hosting environment with the proxy disabled to verify.

### Managing Clients

> Applications should manage how they open and close the eventhub client.  It is essential to be aware that your application is responsible for calling `close()` when it is finished using a client or to use the `with statement` for clients so that they are automatically closed after the flow execution leaves that block. This make sures that the socket connection to the server is closed correctly.

Reviewer 2: ```suggestion
Applications should manage how they open and close the eventhub client.  It is essential to be aware that your application is responsible for calling `close()` when it is finished using a client or to use the `with` statement for clients so that they are automatically closed after the flow execution leaves that block. This make sures that the socket connection to the server is closed correctly.
```

Reviewer 2: may be helpful to add at the end: `If the client is not closed, you may see errors such as..." or other unwanted behavior?

This error can occur when an intercepting proxy is used.  We recommend testing in your hosting environment with the proxy disabled to verify.

### Managing Clients

Applications should manage how they open and close the eventhub client.  It is essential to be aware that your application is responsible for calling `close()` when it is finished using a client or to use the `with` statement for clients so that they are automatically closed after the flow execution leaves that block. This make sures that the socket connection to the server is closed correctly. If the client is not closed this can leave AMQP connections open, sockets open and prevent other clean up activities from running.

> ### Adding components to the connection string does not work

Reviewer 2: ```suggestion
### Cannot add components to the connection string
```

Reviewer 2: This section should be moved below the IoT one to match the order of the list at the top.

The current generation of the Event Hubs client library supports connection strings only in the form published by the Azure portal. These are intended to provide basic location and shared key information only; configuring behavior of the clients is done through its options.

Previous generations of the Event Hub clients allowed for some behavior to be configured by adding key/value components to a connection string. These components are no longer recognized and have no effect on client behavior.

> Because translating a connection string requires querying the IoT Hub service, the Event Hubs client library cannot use it directly.  The [IoT Hub Connection String Sample][IoTConnectionString] sample describes how to query IoT Hub to translate an IoT connection string into one that can be used with Event Hubs.

Reviewer 2: IoT section name should be added back in above IoT paragraph:
```suggestion
### Connect using an IoT connection string

Because translating a connection string requires querying the IoT Hub service, the Event Hubs client library cannot use it directly.  The [IoT Hub Connection String Sample][IoTConnectionString] sample describes how to query IoT Hub to translate an IoT connection string into one that can be used with Event Hubs.
```

#### Adding "TransportType=AmqpWebSockets"

To use web sockets, pass in a kwarg `transport_type = TransportType.AmqpOverWebsocket` during client [creation][WebsocketConfig].

#### Adding "Authentication=Managed Identity"

To authenticate with Managed Identity, see the [sample](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/eventhub/azure-eventhub/samples/sync_samples/client_identity_authentication.py).

For more information about the `azure.identity` library, check out our [Authentication and the Azure SDK][AuthenticationAndTheAzureSDK] blog post.

## Enable and configure logging

The Azure SDK for Python offers a consistent logging story to help troubleshoot application errors and expedite their resolution.  The logs produced will capture the flow of an application before reaching the terminal state to help locate the root issue.

> This library uses the standard [Logging] library for logging and an example to set up debug logging with transport level logging is [here](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/eventhub/azure-eventhub#logging). 

Reviewer 2: We may want to copy that entire section over here in addition to the linking. It'll be easier to copy paste from here.

However, if no instances are being added or removed, there is an underlying issue that should be addressed.  See [Partition ownership changes a lot](#partition-ownership-changes-a-lot) for additional information and [Filing GitHub issues](#filing-github-issues).

### Blocking Calls in Async

> When working with the async client, Applications should ensure that they are not blocking the event loop as the load balancing operation runs as a task in the background. Blocking the event loop can impact load balancing, error handling, check pointing and recovering after an error. CPU bound code that can block the event loop should be run in [executor](https://docs.python.org/3/library/asyncio-dev.html#running-blocking-code)

Reviewer 2: nit
```suggestion
When working with the async client, applications should ensure that they are not blocking the event loop as the load balancing operation runs as a task in the background. Blocking the event loop can impact load balancing, error handling, checkpointing, and recovering after an error. CPU-bound code that can block the event loop should be run in [executor](https://docs.python.org/3/library/asyncio-dev.html#running-blocking-code).
```

However, if no instances are being added or removed, there is an underlying issue that should be addressed.  See [Partition ownership changes a lot](#partition-ownership-changes-a-lot) for additional information and [Filing GitHub issues](#filing-github-issues).

### Blocking Calls in Async

> When working with the async client, applications should ensure that they are not blocking the event loop as the load balancing operation runs as a task in the background. Blocking the event loop can impact load balancing, error handling, checkpointing, and recovering after an error. CPU-bound code that can block the event loop should be run in [executor] [executor](https://docs.python.org/3/library/asyncio-dev.html#running-blocking-code)

Reviewer 2: ```suggestion
When working with the async client, applications should ensure that they are not blocking the event loop as the load balancing operation runs as a task in the background. Blocking the event loop can impact load balancing, error handling, checkpointing, and recovering after an error. CPU-bound code that can block the event loop should be run in [executor](https://docs.python.org/3/library/asyncio-dev.html#running-blocking-code)
```

### Blocking Calls in Async

When working with the async client, applications should ensure that they are not blocking the event loop as the load balancing operation runs as a task in the background. Blocking the event loop can impact load balancing, error handling, checkpointing, and recovering after an error. CPU-bound code that can block the event loop should be run in [executor](https://docs.python.org/3/library/asyncio-dev.html#running-blocking-code)

### High CPU usage

High CPU usage is usually because an instance owns too many partitions.  We recommend no more than three partitions for every 1 CPU core; better to start with 1.5 partitions for each CPU core and test increasing the number of partitions owned.

### One or more partitions have high latency for processing

When processing for one or more partitions is delayed, it is most often because an event processor owns too many partitions. Consider adding additional consumers to handle partitions, we generally recommend 1 - 3 partitions per consumer depending on throughput.

### Processor client stops receiving

> The processor client often is continually running in a host application for days on end.  Sometimes, they notice that EventHubConsumerClient is not processing one or more partitions.  Usually, this is not enough information to determine why the exception occurred.  The EventHubConsumerClient stopping is the symptom of an underlying cause that occurred while trying to recover from a transient error.  Please see [Filing Github issues](#filing-github-issues) for the information we require.

Reviewer: the wording here is a bit confusing -- what are we trying to tell them, if your consumer stops file a bug?

### Duplicate events are being processed

This is usually normal and most often does not indicate an issue unless partitions are frequently closing and initializing.

An important call-out is that Event Hubs has an at-least-once delivery guarantee; it is highly recommended that applications ensure that processing is resilient to event duplication in whatever way is appropriate for their environment and application scenarios.

Event processors configured to use the same Event Hub, consumer group, and checkpoint store will coordinate with one another to share the responsibility of processing partitions. When a processor isn’t able to reach its fair share of the work by claiming unowned partitions, it will attempt to steal ownership from other event processors. During this time, the new owner will begin reading from the last recorded checkpoint. At the same time, the old owner may be dispatching the events that it last read to the handler for processing; it will not understand that ownership has changed until it attempts to read the next set of events from the Event Hubs service.

> As a result, there will be an amount of duplicate events being processed when event processors are started or stopped which will subside when partition ownership has stabilized. If frequent partition initialization persists longer than five minutes, it likely indicates a problem.

Reviewer: frequent partition initialization persists -> if the partition frequently takes 5 minutes to initialize?

However, if no instances are being added or removed, there is an underlying issue that should be addressed.  See [Partition ownership changes a lot](#partition-ownership-changes-a-lot) for additional information and [Filing GitHub issues](#filing-github-issues).

### Blocking Calls in Async

> When working with the async clients, applications should ensure that they are not blocking the event loop as the load balancing operation runs as a task in the background. Blocking the event loop can impact load balancing, error handling, checkpointing, and recovering after an error. CPU-bound code that can block the event loop should be run in [executor](https://docs.python.org/3/library/asyncio-dev.html#running-blocking-code)

Reviewer 2: minor tiny nitty nit
```suggestion
When working with the async clients, applications should ensure that they are not blocking the event loop as the load balancing operation runs as a task in the background. Blocking the event loop can impact load balancing, error handling, checkpointing, and recovering after an error. CPU-bound code that can block the event loop should be run in [executor](https://docs.python.org/3/library/asyncio-dev.html#running-blocking-code).
```

### Blocking Calls in Async

When working with the async client, applications should ensure that they are not blocking the event loop as the load balancing operation runs as a task in the background. Blocking the event loop can impact load balancing, error handling, checkpointing, and recovering after an error. CPU-bound code that can block the event loop should be run in [executor](https://docs.python.org/3/library/asyncio-dev.html#running-blocking-code)

### High CPU usage

High CPU usage is usually because an instance owns too many partitions.  We recommend no more than three partitions for every 1 CPU core; better to start with 1.5 partitions for each CPU core and test increasing the number of partitions owned.

### One or more partitions have high latency for processing

When processing for one or more partitions is delayed, it is most often because an event processor owns too many partitions. Consider adding additional consumers to handle partitions, we generally recommend 1 - 3 partitions per consumer depending on throughput.

### A partition is not being processed

> An event consumer runs continually in a host application for a prolonged period. Sometimes, it may appear that some partitions are uncrowned or are not being processed. Most often, this presents as [Partions close and initialize frequently](#partitions-close-and-initialize-frequently) or there are other errors/warning being raised related to the problem.

Reviewer: typo 
```suggestion
An event consumer runs continually in a host application for a prolonged period. Sometimes, it may appear that some partitions are uncrowned or are not being processed. Most often, this presents as [Partitions close and initialize frequently](#partitions-close-and-initialize-frequently) or there are other errors/warning being raised related to the problem.
```

<!-- repo links -->

> [IoTConnectionString]: https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/eventhub/azure-eventhub/samples/async_samples/iot_hub_connection_string_receive_async.py

Reviewer 2: need to add links for AmqpRetryOptions and ClientCreation

[PublishEventsWithWebSocketsAndProxy]: https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/eventhub/azure-eventhub/samples/async_samples/proxy_async.py

> [WebsocketConfig]: https://github.com/Azure/azure-sdk-for-python/blob/
> ab0ed5efa371db62dd820fdaed70b1e6821f4e0c/sdk/eventhub/azure-eventhub/samples/async_samples/send_and_receive_amqp_annotated_message_async.py#L95

Reviewer 2: ```suggestion
[WebsocketConfig]: https://github.com/Azure/azure-sdk-for-python/blob/ab0ed5efa371db62dd820fdaed70b1e6821f4e0c/sdk/eventhub/azure-eventhub/samples/async_samples/send_and_receive_amqp_annotated_message_async.py#L95
```

* What is the traffic pattern like in your Event Hub?  (i.e. # messages/minute and if the EventHubConsumerClient is always busy or has slow traffic periods.)
* Repro code and steps
  * This is important as we often cannot reproduce the issue in our environment.

> * Logs.  We need DEBUG logs, but if that is not possible, INFO at least.  Error and warning level logs do not provide enough information.  The period of at least +/- 10 minutes from when the issue occurred. See the [Enable and configure logging](#enable-and-configure-logging) section for more information.

Reviewer: can we make this more visible to customers? make when filing github issues its own section?

> <!-- repo links -->

Reviewer 2: Can we add a link next to the logs bullet point above for the logging section in this guide?

When filing GitHub issues, the following details are requested:

* Event Hubs environment
  * How many partitions?
* EventHubConsumerClient environment
  * What is the machine(s) specs processing your Event Hubs?
  * How many instances are running?
* What is the average size of each EventData?
* What is the traffic pattern like in your Event Hubs?  (i.e. # messages/minute and if the EventHubConsumerClient is always busy or has slow traffic periods.)
* Repro code and steps
  * This is important as we often cannot reproduce the issue in our environment.

> * Logs.  DEBUG level logs are needed to best help you. If that is not possible, INFO level logs must be provided.  ERROR and WARNING level logs do not provide enough information to help debug the issue in most cases.  The time period for logst must be at least +/- 10 minutes from when the issue occurred. See the [Enable and configure logging](#enable-and-configure-logging) section for more information.

Reviewer 2: ```suggestion
* Logs.  DEBUG level logs are needed to best help you. If that is not possible, INFO level logs must be provided.  ERROR and WARNING level logs do not provide enough information to help debug the issue in most cases.  The time period for logs must be at least +/- 10 minutes from when the issue occurred. See the [Enable and configure logging](#enable-and-configure-logging) section for more information.
```
